lib/rule-context.js

```diff
@@ -185,6 +185,19 @@ class RuleContext {
             const descriptor = nodeOrDescriptor;
             const fix = getFix(descriptor, this.getSourceCode());
 
+            if (descriptor.messageId) {
+                const id = descriptor.messageId;
+                const messages = this.meta.messages;
+
+                if (descriptor.message) {
+                    throw new TypeError("context.report() called with a message and a messageId. Please only pass one.");
+                }
+                if (!messages || Object.keys(messages).indexOf(id) === -1) {
+                    throw new TypeError(`context.report() called with a messageId of '${id}' which is not present in the 'messages' config: ${JSON.stringify(messages, null, 2)}`);
+                }
+                descriptor.message = messages[id];
```

Comment:
I see we're only using `messageId` in `RuleContext`, and it gets replaced with `message` by the time it reaches formatters. Do we want to preserve `messageId` in formatters? That would allow formatters to do i18n by swapping out messages based on the `messageId`, which I thought was a goal of the proposal.

(That would also make the `messageId` part of a rule's public API since changing it could break external formatters. This may or may not be desirable.)

(Also, it might be difficult for formatters to replace messages without having access to the `data` from the rule, since it won't be possible to unambiguously parse interpolated messages to determine what the interpolated data is.)

Reply:
I was thinking that rules would do something like

```js
messages: {
  foo: _('Don’t do foo!')
}
```